### ISSUE TITLE
Add medm screen viewer

### ISSUE THREAD

**Issue by stefco** [stefco/geco_vm#29]

**Comment by stefco** [comment]
instructions: https://wiki.ligo.org/RemoteAccess/RemoteMEDMBuildFromSource

**Comment by stefco** [comment]
which readline?

```
readline-common - GNU readline and history libraries, common files
lib64readline-gplv2-dev - GNU readline and history libraries, development files (64-bit)
lib64readline5 - GNU readline and history libraries, run-time libraries (64-bit)
lib64readline6 - GNU readline and history libraries, run-time libraries (64-bit)
lib64readline6-dev - GNU readline and history libraries, development files (64-bit)
```

```
libreadline-dev - GNU readline and history libraries, development files
libreadline-gplv2-dev - GNU readline and history libraries, development files
libreadline5 - GNU readline and history libraries, run-time libraries
libreadline5-dbg - GNU readline and history libraries, debugging libraries
libreadline6 - GNU readline and history libraries, run-time libraries
libreadline6-dbg - GNU readline and history libraries, debugging libraries
libreadline6-dev - GNU readline and history libraries, development files
```

**Comment by stefco** [comment]
try libreadline6 and libreadline6-dev http://askubuntu.com/questions/194523/how-do-i-install-gnu-readline

**Comment by stefco** [comment]
- [x] install everything
- [x] see if any necessary x11 packages get clobbered by cleanup scripts
- [x] if so, see what happens when they get clobbered
- [x] if clobbering breaks everything, then do not clobber

**Comment by stefco** [comment]
download epics source from here http://www.aps.anl.gov/epics/download/index.php

**Comment by stefco** [comment]
Alarm handler not building:

```
/usr/bin/gcc -c  -D_GNU_SOURCE -D_DEFAULT_SOURCE            -D_X86_64_  -DUNIX  -Dlinux     -O3 -g   -Wall       -m64      -MMD -I. -I../O.Common -I. -I.. -I../../../include/os/Linux -I../../../include -I../../../../base/include/os/Linux -I../../../../base/include        -I/usr/include -I/usr/include    ../productDescriptionShell.c
make[1]: *** No rule to make target `/usr/lib64/libXm.a', needed by `probe'.  Stop.
make[1]: Leaving directory `/opt/epics/extensions/src/probe1_1_8_0/O.linux-x86_64'
make: *** [install.linux-x86_64] Error 2
root@vagrant:/vagrant/scripts/geco# echo 'cd /opt/epics/extensions/src/alh1_2_35; make' | bash
make -C O.linux-x86_64 -f ../Makefile TOP=../../.. \
            T_A=linux-x86_64 install
make[1]: Entering directory `/opt/epics/extensions/src/alh1_2_35/O.linux-x86_64'
make[1]: *** No rule to make target `/usr/lib64/libXmu.a', needed by `alh'.  Stop.
make[1]: Leaving directory `/opt/epics/extensions/src/alh1_2_35/O.linux-x86_64'
make: *** [install.linux-x86_64] Error 2
```

emphasis: `No rule to make target '/usr/lib64/libXmu.a', needed by 'alh'.  Stop.`

see:
http://www.aps.anl.gov/epics/tech-talk/2007/msg00935.php
http://www.aps.anl.gov/epics/base/R3-14/12-docs/ConvertingR3.13AppsToR3.14.html

Probe not building:

```
root@vagrant:/vagrant/scripts/geco# echo 'cd /opt/epics/extensions/src/probe1_1_8_0; make' | bash
make -C O.linux-x86_64 -f ../Makefile TOP=../../.. \
            T_A=linux-x86_64 install
make[1]: Entering directory `/opt/epics/extensions/src/probe1_1_8_0/O.linux-x86_64'
make[1]: *** No rule to make target `/usr/lib64/libXm.a', needed by `probe'.  Stop.
make[1]: Leaving directory `/opt/epics/extensions/src/probe1_1_8_0/O.linux-x86_64'
make: *** [install.linux-x86_64] Error 2
```

emphasis: `No rule to make target '/usr/lib64/libXm.a', needed by 'probe'.  Stop.`

see:
http://www.aps.anl.gov/epics/tech-talk/2013/msg00347.php

**Comment by stefco** [comment]
x11 libs seem necessary; keep them all for now

**Comment by stefco** [comment]
ugh, MEDM also not building:

```
/usr/bin/gcc -c  -D_GNU_SOURCE -D_DEFAULT_SOURCE            -D_X86_64_  -DUNIX  -Dlinux     -O3 -g   -Wall     -DSCIPLOT -DMOTIF -DEDITRES   -m64      -MMD -I. -I../O.Common -I. -I.. -I../../../../include/os/Linux -I../../../../include -I../../../../../base/include/os/Linux -I../../../../../base/include        -I../../printUtils -I../../xc -I/usr/include -I/usr/include    ../xgif.c
../xgif.c: In function ‘initializeGIF’:
../xgif.c:157:21: warning: variable ‘h’ set but not used [-Wunused-but-set-variable]
../xgif.c:157:18: warning: variable ‘w’ set but not used [-Wunused-but-set-variable]
../xgif.c:156:12: warning: variable ‘y’ set but not used [-Wunused-but-set-variable]
../xgif.c:156:9: warning: variable ‘x’ set but not used [-Wunused-but-set-variable]
make[2]: *** No rule to make target `/usr/lib64/libXm.a', needed by `medm'.  Stop.
make[2]: Leaving directory `/opt/epics/extensions/src/medm3_1_9/medm/O.linux-x86_64'
make[1]: *** [install.linux-x86_64] Error 2
make[1]: Leaving directory `/opt/epics/extensions/src/medm3_1_9/medm'
make: *** [medm.install] Error 2
```

emphasis: `No rule to make target '/usr/lib64/libXm.a', needed by 'medm'.  Stop.`

**Comment by stefco** [comment]
apparent solution: install [MEDM dependencies](http://www.aps.anl.gov/epics/extensions/medm/index.php) and [change CONFIG_SITE](http://www.aps.anl.gov/epics/tech-talk/2015/msg00091.php) to reflect the actual location of `libXm.a`

attempting now

**Comment by stefco** [comment]
```
root@vagrant:/vagrant/scripts/geco# find / -name libXm.a
/usr/lib/x86_64-linux-gnu/libXm.a
```

```
root@vagrant:/vagrant/scripts/geco# find / -name libXmu.a
/usr/lib/x86_64-linux-gnu/libXmu.a
```

**Comment by stefco** [comment]
looks like i gotta modify linux-x86_64 versions of the CONFIG_SITE files:

```
root@vagrant:/opt/epics/extensions/configure/os# grep 'usr/lib64' *
CONFIG_SITE.linux-x86_64.linux-x86_64:X11_LIB=/usr/lib64
CONFIG_SITE.linux-x86_64.linux-x86_64:MOTIF_LIB=/usr/lib64
```

**Comment by stefco** [comment]
how to make a patch file: https://docs.moodle.org/dev/How_to_create_a_patch

my patch command:

``` bash
root@vagrant:/opt/epics/extensions/configure/os# diff -Naur \
    CONFIG_SITE.linux-x86_64.linux-x86_64 \
    CONFIG_SITE.linux-x86_64.linux-x86_64.fixed
```

output:

```
--- CONFIG_SITE.linux-x86_64.linux-x86_64       2009-06-04 14:22:28.000000000 +0000
+++ CONFIG_SITE.linux-x86_64.linux-x86_64.fixed 2016-03-23 21:30:46.596537927 +0000
@@ -11,11 +11,11 @@

 -include $(TOP)/configure/os/CONFIG_SITE.linux-x86.linux-x86

-X11_LIB=/usr/lib64
+X11_LIB=/usr/lib/x86_64-linux-gnu
 X11_INC=/usr/include

 # OpenMotif location
-MOTIF_LIB=/usr/lib64
+MOTIF_LIB=/usr/lib/x86_64-linux-gnu
 MOTIF_INC=/usr/include
 #MOTIF_LIB=/usr/X11R6/lib64
 #MOTIF_INC=/usr/X11R6/include
root@vagrant:/opt/epics/extensions/configure/os#
```

save the diffs to a file, which can then be used by `patch`:

```
root@vagrant:/opt/epics/extensions/configure/os# diff -Naur \
    CONFIG_SITE.linux-x86_64.linux-x86_64 \
    CONFIG_SITE.linux-x86_64.linux-x86_64.fixed > ubuntu-lib-locations.diff
```

I should have run it in `/opt/epics/extensions/` so that the path would start from there, mimicking the patches provided by Jon Hanks. Manually edited it by adding `configure/os/` to the beginning of the paths showing on the first two lines of the diff file, giving

```
--- configure/os/CONFIG_SITE.linux-x86_64.linux-x86_64       2009-06-04 14:22:28.000000000 +0000
+++ configure/os/CONFIG_SITE.linux-x86_64.linux-x86_64.fixed 2016-03-23 21:30:46.596537927 +0000
```

which makes it possible to run the patch from `/opt/epics/extensions/`, just like in the instructions.

**Comment by stefco** [comment]
running through a full test with

``` bash
rm -rf /opt/epics && cp -R /vagrant/http/epics /opt && sh -e /vagrant/scripts/geco/install-medm.sh
```

**Comment by stefco** [comment]
hunks failing during patch!

```
patching file medm/Makefile
patching file medm/cache.c
patching file medm/cache.h
patching file medm/display.c
patching file medm/medm.c
patching file medm/medmCommon.c
patching file medm/medmComposite.c
patching file medm/medmRelatedDisplay.c
patching file medm/medmWidget.h
patching file medm/membuffer.c
patching file medm/membuffer.h
patching file medm/network.c
patching file medm/proto.h
patching file medm/utils.c
Hunk #1 succeeded at 43 with fuzz 2 (offset -1 lines).
Hunk #2 succeeded at 137 (offset -1 lines).
Hunk #3 succeeded at 159 (offset -1 lines).
Hunk #4 FAILED at 3175.
Hunk #5 FAILED at 3198.
2 out of 5 hunks FAILED -- saving rejects to file medm/utils.c.rej
patching file medm/xgif.c
```

**Comment by stefco** [comment]
detailed output in `medm/utils.c.rej`:

```
--- medm/utils.c
+++ medm/utils.c
@@ -3175,6 +3182,7 @@
  *  A. Ivanov, K. Thorne, January, 2012 LIGO 
  */
 void read_macros(char *str, char *buf) {
+  char fullPathName[MAX_TOKEN_LENGTH];
   char *pch = strtok (str,",");
   *buf = 0;
   while (pch != NULL)
@@ -3198,29 +3206,36 @@
                         if (argv[1] == 0) {
                                  medmPrintf(1, "Unspecified file name in macro read command\n");
                         } else {
-               char *fname=STRDUP(argv[1]);
-                                FILE *fp = fopen(fname, "r");
-                                if (fp == 0) {
-                                        medmPrintf(1, "Cannot open macro file %s\n", fname);
-                                } else {
-                                        struct stat st;
-                                        char buffer[MAX_TOKEN_LENGTH];
-                                        fstat(fileno(fp), &st);
-                   size_t bytes_read;
-                                        if (st.st_size > (MAX_TOKEN_LENGTH-1)) {
-                                                medmPrintf(1, "File %s too big\n", fname);
-                                        } else {       
-                                           bytes_read = fread(buffer, sizeof(char), st.st_size, fp);
-                                                buffer[st.st_size] = 0;
-                                               /* Trim trailing space */
-                                                char *end = buffer + strlen(buffer) - 1;
-                                                while(end > buffer && isspace(*end)) end--;
-                                                /* Write new null terminator */
-                                                *(end+1) = 0;
-                                                strcat(buf, buffer);
-                                        }
-                                        fclose(fp);
-                                }
+                               char *fname=STRDUP(argv[1]);
+                               if (!locateResource(fullPathName, fname, MAX_TOKEN_LENGTH)) {
+                                   medmPrintf(1, "Cannot locate macro file %s\n", fname);
+                               } else {
+                                   MemBuffer *m = NULL;
+                                   if (isNetworkRequest(fullPathName)) {
+                                       m = getResourceFromURL(fullPathName);
+                                   } else {
+                                       m = mbNewFromFile(fullPathName, "rt");
+                                   }
+                                   if (!m) {
+                                       medmPrintf(1, "Unable to load macro file %s\n", fullPathName);
+                                   } else {
+                                       if ((m->length + strlen(buf)) > MAX_TOKEN_LENGTH-1) {
+                                           medmPrintf(1, "File %s too big\n", fullPathName);
+                                       } else {
+                                           char *end = buf;
+                                           while (*end != '\0') {
+                                               end++;
+                                           }
+                                           memcpy(end, m->data, m->length);
+                                           end += m->length;
+                                           *end = '\0';
+                                           end--;
+                                           while (end > buf && isspace(*end)) end--;
+                                           *(end+1) = '\0';
+                                       }
+                                       m = mbClose(m);
+                                   }
+                           }
                         }
                 } else {
                         medmPrintf(1, "Unknown macro command %s\n", argv[0]);
```

**Comment by stefco** [comment]
this junk in the previous:

```
  *  A. Ivanov, K. Thorne, January, 2012 LIGO 
  */
```

suggests that there is a missing patch.

**Comment by stefco** [comment]
downloaded this little guy: https://www.ligo-wa.caltech.edu/~jonathan.hanks/epics/tgz/epics_remote_medm-2014-11-24.tar.gz

and ran:

```
grep -rn 'K. Thorne, January, 2012 LIGO' .
```

returning:

```
./patches/extensions/medm3_1_7_macrofile.patch
```

likewise,

```
grep -Ern '^---.*utils\.c' .
```

returned

```
./patches/extensions/medm3_1_7_macrofile.patch
```

suggesting that this might be the place where the necessary prior patches are introduced. hopefully.
- [ ] Contact Jon Hanks about this!!

**Comment by stefco** [comment]
Full text of above file:

```
--- medm3_1_7/medm/utils.c  2012-06-21 15:03:01.000000000 -0500
+++ medm3_1_7/medm/utils.c  2013-03-07 15:47:14.453023205 -0600
@@ -42,6 +42,7 @@
 #include <cvtFast.h>
 #include <postfix.h>
 #include "medm.h"
+#include <sys/stat.h>

 /* Include this after medm.h to avoid problems with Exceed 6 */
 #ifdef WIN32
@@ -3162,6 +3163,69 @@
 }

 /*** Name value table routines ***/
+/* 
+ * read_macros - allow %(read macro_file) option 
+ *  A. Ivanov, K. Thorne, January, 2012 LIGO 
+ */
+void read_macros(char *str, char *buf) {
+  char *pch = strtok (str,",");
+  *buf = 0;
+  while (pch != NULL)
+  {
+    if (*pch == '%' && pch[1] == '(') {
+        /* This is the directive to do stuff */
+        /* skip over "$(" */
+        pch+=2;  
+        /* find closing bracket */
+        char *endbrkt = strstr(pch,")");
+        if (endbrkt) {
+                *endbrkt = 0;
+                /* Split the string into two fields */
+                char **ap, *argv[2];
+                for (ap = argv; (*ap = strsep(&pch, " \t")) != NULL;)
+                   if (**ap != '\0')
+                           if (++ap >= &argv[2])
+                                   break;
+                if (!strcmp("read", argv[0])) {
+                        /* Open the file */
+                        if (argv[1] == 0) {
+                                 medmPrintf(1, "Unspecified file name in macro read command\n");
+                        } else {
+               char *fname=STRDUP(argv[1]);
+                                FILE *fp = fopen(fname, "r");
+                                if (fp == 0) {
+                                        medmPrintf(1, "Cannot open macro file %s\n", fname);
+                                } else {
+                                        struct stat st;
+                                        char buffer[MAX_TOKEN_LENGTH];
+                                        fstat(fileno(fp), &st);
+                   size_t bytes_read;
+                                        if (st.st_size > (MAX_TOKEN_LENGTH-1)) {
+                                                medmPrintf(1, "File %s too big\n", fname);
+                                        } else {       
+                                           bytes_read = fread(buffer, sizeof(char), st.st_size, fp);
+                                                buffer[st.st_size] = 0;
+                                               /* Trim trailing space */
+                                                char *end = buffer + strlen(buffer) - 1;
+                                                while(end > buffer && isspace(*end)) end--;
+                                                /* Write new null terminator */
+                                                *(end+1) = 0;
+                                                strcat(buf, buffer);
+                                        }
+                                        fclose(fp);
+                                }
+                        }
+                } else {
+                        medmPrintf(1, "Unknown macro command %s\n", argv[0]);
+                }
+        }
+    } else {
+        strcat(buf, pch);
+    }
+    pch = strtok (NULL, ",");
+    if (pch) strcat(buf, ",");
+  }
+}

 /*
  * generate a name-value table from the passed-in argument string
@@ -3171,7 +3235,7 @@
  */
 NameValueTable *generateNameValueTable(char *argsString, int *numNameValues)
 {
-    char *copyOfArgsString,  *name, *value;
+    char *name, *value;
     char *s1;
     char nameEntry[80], valueEntry[80];
     int i = 0, j = 0, tableIndex = 0, numPairs = 0, numEntries = 0;
@@ -3180,11 +3244,13 @@


     nameTable = NULL;
-    copyOfArgsString = NULL;

     if(argsString != NULL) {

-   copyOfArgsString = STRDUP(argsString);
+   char copyOfArgsString[MAX_TOKEN_LENGTH];
+   char *strcp = STRDUP(argsString);
+   read_macros(strcp, copyOfArgsString);
+   free(strcp);
       /* see how many a=b name/value pairs are in the string */
    numPairs = 0;
    i = 0;
@@ -3228,7 +3294,6 @@
        tableIndex++;
        }
    }
-   if(copyOfArgsString) free(copyOfArgsString);

     } else {

```

**Comment by stefco** [comment]
aoj4pt39tjcp9i of;kfioj;kmlL:@#<LFGER

SUCCESS.

**Comment by stefco** [comment]
- [x] add hard link to `ubuntu-lib-locations.diff` to the script, so that other users can use it as well

**Comment by stefco** [comment]
having trouble, suspect it might be related to port, since i need to use port forwarding for this:

```
vagrant@vagrant:~$ medm_lho -u stefan.countryman
Args =  after parsing
SOCKDIR=/tmp/epics.yV6WN9XcDfWt
stefan.countryman@lhoepics.ligo-wa.caltech.edu's password:
Connection to lhoepics.ligo-wa.caltech.edu established
Searching for a free port to use for EPICS CA transport
The script will randomly select some ports in an attempt to find
an available network port to send the EPICS data over.
Attempting to use port 15388 for EPICS
Forward attempt 1
arg: /ligo/cds/lho/medm/sitemap.adl
locateResource '/ligo/cds/lho/medm/sitemap.adl'
isNetworkRequest /ligo/cds/lho/medm/sitemap.adl
canAccess('/ligo/cds/lho/medm/sitemap.adl', 4) = -1
searching for '/ligo/cds/lho/medm/sitemap.adl' in 'https://lhoepics.ligo-wa.caltech.edu/'
isNetworkRequest https://lhoepics.ligo-wa.caltech.edu///ligo/cds/lho/medm/sitemap.adl
cmdBuffer = "curl -sw '%{http_code}' https://lhoepics.ligo-wa.caltech.edu///ligo/cds/lho/medm/sitemap.adl"
newMemBufferFromURL(https://lhoepics.ligo-wa.caltech.edu///ligo/cds/lho/medm/sitemap.adl) returned success with a length of 46492
canAccess('https://lhoepics.ligo-wa.caltech.edu///ligo/cds/lho/medm/sitemap.adl', 4) = 0
locateResource(/ligo/cds/lho/medm/sitemap.adl...) returning 1
Error: Can't open display:
Shutting down encrypted tunnel
Exit request sent.
Removing temporary directories
```

**Comment by stefco** [comment]
tried running `xarclock` as a simple test of x window forwarding, and this also didn't work. could be due to uninstallation of x window packages during cleanup. am rebuilding the box now to bake in the changes and will resume debuggin x window forwarding from here.
- [x] debug x window
- [x] see if EPICS's erratic port usage is part of the problem (seems it would be)

**Comment by stefco** [comment]
could also forget about it and install a gui; would make life easier for people who do not want to install x11 clients on their macs and pcs at the expense of virtual image size and suitability as a server.

**Comment by stefco** [comment]
instructions for installing mate:

http://wiki.mate-desktop.org/download

**Comment by stefco** [comment]
http://ubuntuforums.org/archive/index.php/t-2035830.html has **NON broken instructions...**

one of them suggested installing the following for virtualbox x window forwarding:

``` bash
apt-get install virtualbox-guest-x11
```

installing that generated:

```
Suggested packages:
  libglide3 xfs xserver xfonts-scalable
```

**Comment by stefco** [comment]
x11 was not installed on this mac... there was just a dummy icon. thank you apple.

**Comment by stefco** [comment]
xarclock is not working on `hashicorp/precise64` either for some reason...

**Comment by stefco** [comment]
xarclock does work on home computer, probably just need to restart the other computer or something. meanwhile libraries are indeed uninstalled during the cleanup process, which is not good.

**Comment by stefco** [comment]
Should put this all in a separate script to avoid security concerns and add in uname/password auth for downloads
- [ ] download straight from jonathan hanks ligo pages
- [ ] possibly cat the patch straight into opt/epics/distfiles from a heredoc to make install more self-contained

**Comment by stefco** [comment]
xarclock is working on computer after restart, time to check `medm`

**Comment by stefco** [comment]
medm works!!!

**Comment by stefco** [comment]
there are occasional button placement/click detection and window position issues when running MEDM on mac. sometimes the window will appear offscreen. for `medm_llo`, the window will be placed off screen on startup.

both issues can be fixed by clicking `Window > Zoom` twice to zoom and unzoom the window.

additionally, the windows sometimes have trouble getting information from the servers. closing and reopening them usually fixes this.

**Comment by stefco** [comment]
also, port forwarding is actually not an issue; the port specified is probably a remote port, not a port on geco-vm that needs to be opened.

**Comment by stefco** [comment]
fixed in #41
